File: packages/sdk-ts/src/core/wire.ts
/**
 * Protobuf wire primitives, written by hand.
 *
 * The canonical encoder cannot delegate to a general-purpose protobuf library: those
 * libraries legitimately preserve unknown fields, tolerate multiple valid orderings, and
 * differ on whether an explicitly-set zero is emitted. Every one of those degrees of
 * freedom is a second accepted form, and EN-02 permits exactly one.
 *
 * Specification: Plans/02-CONTRACTS-CORE.md §1
 */

export const WIRE_VARINT = 0;
export const WIRE_LEN = 2;

/** Growable byte sink. Avoids O(n²) concatenation while encoding. */
export class ByteWriter {
  #buf: Uint8Array;
  #len = 0;

  constructor(initialCapacity = 256) {
    this.#buf = new Uint8Array(initialCapacity);
  }

  #ensure(extra: number): void {
    const needed = this.#len + extra;
    if (needed <= this.#buf.length) return;
    let cap = this.#buf.length * 2;
    while (cap < needed) cap *= 2;
    const next = new Uint8Array(cap);
    next.set(this.#buf.subarray(0, this.#len));
    this.#buf = next;
  }

  byte(b: number): void {
    this.#ensure(1);
    this.#buf[this.#len++] = b;
  }

  bytes(src: Uint8Array): void {
    this.#ensure(src.length);
    this.#buf.set(src, this.#len);
    this.#len += src.length;
  }

  /** Base-128 varint, little-endian groups, high bit as continuation. */
  varint(value: bigint): void {
    let v = value;
    if (v < 0n) {
      // Two's complement over 64 bits — protobuf encodes a negative int64 as ten bytes.
      v = (1n << 64n) + v;
    }
    while (v >= 0x80n) {
      this.byte(Number((v & 0x7fn) | 0x80n));
      v >>= 7n;
    }
    this.byte(Number(v));
  }

  tag(fieldNumber: number, wireType: number): void {
    this.varint(BigInt((fieldNumber << 3) | wireType));
  }

  finish(): Uint8Array {
    return this.#buf.slice(0, this.#len);
  }

  get length(): number {
    return this.#len;
  }
}

/**
 * NFC normalisation before encoding (EN-01).
 *
 * Two visually identical strings can have different UTF-8 bytes — "é" is either U+00E9
 * or U+0065 U+0301. Without normalisation the same text signed on two devices produces
 * two different content IDs, which breaks dedupe, replay protection, and cross-node
 * references all at once.
 */
export function encodeUtf8Nfc(s: string): Uint8Array {
  return new TextEncoder().encode(s.normalize('NFC'));
}

/** Writes `tag, length, payload`. Used for string, bytes, and nested message fields. */
export function writeLengthDelimited(w: ByteWriter, fieldNumber: number, payload: Uint8Array): void {
  w.tag(fieldNumber, WIRE_LEN);
  w.varint(BigInt(payload.length));
  w.bytes(payload);
}

export function writeVarintField(w: ByteWriter, fieldNumber: number, value: bigint): void {
  w.tag(fieldNumber, WIRE_VARINT);
  w.varint(value);
}

/**
 * Length-prefixed concatenation, for signing structures that are NOT envelopes.
 *
 * Receipts, tree heads and the federation handshake payloads all need a deterministic
 * byte string to sign, but none of them is an `Envelope`, so none goes through the
 * canonical encoder. Plain concatenation would be ambiguous — `["ab","c"]` and
 * `["a","bc"]` produce identical bytes — and an attacker who controls two adjacent
 * fields can exploit exactly that to make one signed statement read as another.
 * A four-byte big-endian length before every part removes the ambiguity.
 *
 * Nested repeated fields frame their elements with this same function and contribute the
 * result as a single part, so structure is preserved to any depth.
 */
export function frameParts(parts: readonly Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((size, part) => size + 4 + part.length, 0));
  const view = new DataView(out.buffer);
  let offset = 0;
  for (const part of parts) {
    view.setUint32(offset, part.length, false);
    offset += 4;
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

/** Constant-time-ish equality for byte arrays. Used on signature and hash comparisons. */
export function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= (a[i] as number) ^ (b[i] as number);
  return diff === 0;
}

File: crates/jb-core/src/wire.rs
//! Protobuf wire primitives, hand-written.
//!
//! A general-purpose protobuf library legitimately preserves unknown fields, tolerates
//! several valid orderings, and differs on whether an explicitly-set zero is emitted.
//! Each of those is a second accepted form, and EN-02 permits exactly one.

use unicode_normalization::UnicodeNormalization;

pub const WIRE_VARINT: u32 = 0;
pub const WIRE_LEN: u32 = 2;

/// Base-128 varint. A negative i64 is encoded as its two's-complement u64, which is ten
/// bytes — matching protobuf's int64 behaviour exactly.
pub fn write_varint(out: &mut Vec<u8>, value: u64) {
    let mut v = value;
    while v >= 0x80 {
        out.push(((v & 0x7f) | 0x80) as u8);
        v >>= 7;
    }
    out.push(v as u8);
}

pub fn write_tag(out: &mut Vec<u8>, field_number: u32, wire_type: u32) {
    write_varint(out, u64::from((field_number << 3) | wire_type));
}

pub fn write_varint_field(out: &mut Vec<u8>, field_number: u32, value: u64) {
    write_tag(out, field_number, WIRE_VARINT);
    write_varint(out, value);
}

pub fn write_len_delimited(out: &mut Vec<u8>, field_number: u32, payload: &[u8]) {
    write_tag(out, field_number, WIRE_LEN);
    write_varint(out, payload.len() as u64);
    out.extend_from_slice(payload);
}

/// NFC normalisation before encoding (EN-01).
///
/// Two visually identical strings can have different UTF-8 bytes — "é" is either U+00E9
/// or U+0065 U+0301. Without normalisation the same text signed on two devices produces
/// two different content IDs, breaking dedupe, replay protection and cross-node
/// references at once.
///
/// This must handle Bangla correctly, not merely ASCII: NFR-A04 makes Bangla a
/// first-class locale, broadcast headline budgets are computed against Bangla UTF-8
/// worst case (SIG-26), and a normalisation divergence that only appears in the primary
/// language of the target users is exactly the bug this gate exists to catch.
pub fn encode_utf8_nfc(s: &str) -> Vec<u8> {
    s.nfc().collect::<String>().into_bytes()
}
